File: HackerRank/30 Days of Code/Day12_Inheritance.java
// Day 12: Inheritance
// https://www.hackerrank.com/challenges/30-inheritance/problem

import java.util.*;

class Person {
	protected String firstName;
	protected String lastName;
	protected int idNumber;
	
	// Constructor
	Person(String firstName, String lastName, int identification){
		this.firstName = firstName;
		this.lastName = lastName;
		this.idNumber = identification;
	}
	
	// Print person data
	public void printPerson(){
		 System.out.println(
				"Name: " + lastName + ", " + firstName 
			+ 	"\nID: " + idNumber); 
	}
	 
}

class Student extends Person{
	private int[] testScores;
    
    Student(String f, String l, int i, int[] s){
        super(f, l, i);
        this.testScores = s;
    }
    public char calculate(){
        int avg = 0;
        for (int i : testScores)
            avg += i;
        avg /= testScores.length;
        if (avg >= 90)
            return 'O';
        if (avg >= 80)
            return 'E';
        if (avg >= 70)
            return 'A';
        if (avg >= 55)
            return 'P';
        if (avg >= 40)
            return 'D';
        return 'T';  
    }
}

class Solution {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		String firstName = scan.next();
		String lastName = scan.next();
		int id = scan.nextInt();
		int numScores = scan.nextInt();
		int[] testScores = new int[numScores];
		for(int i = 0; i < numScores; i++){
			testScores[i] = scan.nextInt();
		}
		scan.close();
		
		Student s = new Student(firstName, lastName, id, testScores);
		s.printPerson();
		System.out.println("Grade: " + s.calculate() );
	}
}

// 미리 만들어진 Person 클래스를 상속받아 scores 멤버 배열을 만들고, 해당 값의 평균으로 성적 출력하는 함수 만들기
// C#에서 파생 클래스 생성 및 생성자 조작

File: HackerRank/30 Days of Code/Day12_Inheritance.cpp
// Day 12: Inheritance
// https://www.hackerrank.com/challenges/30-inheritance/problem

#include <iostream>
#include <vector>

using namespace std;

class Person{
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification){
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout<< "Name: "<< lastName << ", "<< firstName <<"\nID: "<< id << "\n"; 
		}
	
};

class Student :  public Person{
	private:
		vector<int> testScores;  
	public:
        Student(string f, string l, int i, vector<int> s) : Person(f, l, i){
            testScores = s;
        }
        char calculate(){
            int avg = 0;
            for (int i : testScores)
                avg += i;
            avg /= testScores.size();
            if (avg >= 90)
                return 'O';
            if (avg >= 80)
                return 'E';
            if (avg >= 70)
                return 'A';
            if (avg >= 55)
                return 'P';
            if (avg >= 40)
                return 'D';
            return 'T';
        }
};

int main() {
	string firstName;
  	string lastName;
	int id;
  	int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  	vector<int> scores;
  	for(int i = 0; i < numScores; i++){
	  	int tmpScore;
	  	cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}

// 미리 만들어진 Person 클래스를 상속받아 scores 멤버 배열을 만들고, 해당 값의 평균으로 성적 출력하는 함수 만들기
// C++에서 파생 클래스 생성 및 생성자 조작